<a href="https://colab.research.google.com/github/MariaMora2723/PRACTICAS-S3-Y-S4/blob/main/Copia_de_TADM2022_S03b_Consultas_SQL_(Clase).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Consultas SQL


Antes de empezar cargamos nuestro archivo data_berka en nuestro entorno: [enlace](https://sorry.vse.cz/~berka/challenge/pkdd1999/berka.htm)

In [ ]:
!gdown --id 1Osmr7qVN0NMuqYxvLesNn7H4nqPCVbsU

/usr/local/lib/python3.7/dist-packages/gdown/cli.py:131: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  category=FutureWarning,
Downloading...
From: https://drive.google.com/uc?id=1Osmr7qVN0NMuqYxvLesNn7H4nqPCVbsU
To: /content/data_berka.db
100% 70.2M/70.2M [00:00<00:00, 162MB/s]


En primer lugar cargamos la extensión que nos va permitir manejar comandos sql directamente en colab. 




In [ ]:
%load_ext sql

Y conectamos con la base de datos, con la siguiente línea se conecta a la base de datos. Para indicar que es código sql, hay que poner al principio siempre %sql

In [ ]:
%sql sqlite:///data_berka.db

'Connected: @data_berka.db'

Comprobamos esquema

Con el siguiente comando le indico que me diga las tablas que hay en la base de datos. 

Con el SELECT le indico la columna que quiero que me coja, con el FROM de dónde y con el WHERE el tipo. 

In [ ]:
%%sql
SELECT name,type
FROM sqlite_master
WHERE type = 'table'

 * sqlite:///data_berka.db
Done.


[('DISTRICT', 'table'),
 ('ACCOUNT', 'table'),
 ('TRANS', 'table'),
 ('CLIENT', 'table'),
 ('DISPOSITION', 'table'),
 ('CREDIT_CARD', 'table'),
 ('LOAN', 'table'),
 ('PERMANENT_ORDER', 'table')]

Al aplicar el comando sabemos que en la base de datos hay 8 tablas. 


# **Consultas sobre Préstamos**

No es lo mismo cuando usamos 1 porcentaje o 2 porcentajes. 
Cuando usamos 1 es para decirle que una lína es SQL y cuando usamos 2 es para indicarle que todo el bloque es SQL. 

1. Obtener todos los atributos que tiene el banco sobre sus prestamos 

(extra limitar a 20 registros)

In [ ]:
%%sql
SELECT *
FROM LOAN
LIMIT 20

 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (6111, 5428, 930924, 174744, 24, 7281, 'B'),
 (7235, 10973, 931013, 154416, 48, 3217, 'A'),
 (5997, 4894, 931104, 117024, 24, 4876, 'A'),
 (7121, 10364, 931110, 21924, 36, 609, 'A'),
 (6077, 5270, 931122, 79608, 24, 3317, 'A'),
 (6228, 6034, 931201, 464520, 60, 7742, 'B'),
 (6356, 6701, 931208, 95400, 36, 2650, 'A'),
 (5523, 2705, 931208, 93888, 36, 2608, 'A'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7104, 10320, 931213, 259740, 60, 4329, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (7097, 10266, 931223, 75624, 24, 3151, 'B')]

2. Obtener los campos de fecha (date), cantidad (amount), duración (duration), pagos (payments), estado (status)

(extra limitar a 20 registros)

En el siguiente comando le he indicado que todo el bloque es sql (con los dos %%), después que me seleccione los campos que me vienen en el enunciado de la tabla LOAN, que me los ordene por cantidad en orden descendiente y que solamente me refleje 20. 

In [ ]:
%%sql 
SELECT date, amount, duration, payments, status
FROM LOAN
ORDER BY amount DESC
LIMIT 20


 * sqlite:///data_berka.db
Done.


[(971019, 590820, 60, 9847, 'C'),
 (980123, 566640, 60, 9444, 'C'),
 (971112, 541200, 60, 9020, 'D'),
 (950217, 538500, 60, 8975, 'C'),
 (980120, 504000, 60, 8400, 'C'),
 (980522, 495180, 60, 8253, 'C'),
 (941219, 482940, 60, 8049, 'D'),
 (970212, 475680, 48, 9910, 'C'),
 (970907, 473280, 60, 7888, 'D'),
 (971225, 468060, 60, 7801, 'C'),
 (970613, 466608, 48, 9721, 'C'),
 (971227, 465504, 48, 9698, 'D'),
 (950709, 465072, 48, 9689, 'C'),
 (931201, 464520, 60, 7742, 'B'),
 (971105, 460980, 60, 7683, 'D'),
 (960723, 444864, 48, 9268, 'C'),
 (980829, 437460, 60, 7291, 'C'),
 (980213, 428784, 48, 8933, 'C'),
 (961118, 421008, 48, 8771, 'C'),
 (951029, 419880, 60, 6998, 'C')]

3. Obtener los distintos tipos de estado que se presentan en un préstamo.

Le he pedido que me indique cuales son los tipos de estados que existen dentro de prestamo de la tabla LOAN, con la palabra "DISTINCT" le indico que me distinga los tipos de estados y me quite los repetidos. 

In [ ]:
%%sql
SELECT  DISTINCT status
FROM LOAN 
ORDER BY status

 * sqlite:///data_berka.db
Done.


[('A',), ('B',), ('C',), ('D',)]

4. Obtener solo los prestamos ya terminados.

Con la línea de where, le indico que me refleje solamente los que dentro de estado esten incluidos como A y B que son los que en los datos nos indican que están ya terminados. 

In [ ]:
%%sql 
SELECT *
FROM LOAN
WHERE status in ("A", "B")
LIMIT 100

 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (6111, 5428, 930924, 174744, 24, 7281, 'B'),
 (7235, 10973, 931013, 154416, 48, 3217, 'A'),
 (5997, 4894, 931104, 117024, 24, 4876, 'A'),
 (7121, 10364, 931110, 21924, 36, 609, 'A'),
 (6077, 5270, 931122, 79608, 24, 3317, 'A'),
 (6228, 6034, 931201, 464520, 60, 7742, 'B'),
 (6356, 6701, 931208, 95400, 36, 2650, 'A'),
 (5523, 2705, 931208, 93888, 36, 2608, 'A'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7104, 10320, 931213, 259740, 60, 4329, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (7097, 10266, 931223, 75624, 24, 3151, 'B'),
 (4959, 2, 940105, 80952, 24, 3373, 'A'),
 (6499, 7401, 940105, 80952, 24, 3373, 'A'),
 (5479, 2486, 940110, 24516, 12, 2043, 'A'),
 (6721, 8475, 940121, 21072, 24, 878, 'A'),
 (7259, 11111, 940131, 108144, 36, 3004, 'A'),
 (5285, 1603, 940206, 78936, 12, 6578, 'A'),
 (5189, 1166, 940207, 149040, 48, 3105, 'A'),
 (6650, 8073, 940217, 49320, 12, 4110, 'B'),
 (7226, 10940, 940223, 197748, 36, 5493, 'A'),
 (7262, 11135, 940301, 182628, 36, 5073, 'A'),
 (5519, 2694, 940308, 91632, 12, 7636, 'A'),
 (6577, 7753, 940311, 51696, 24, 2154, 'A'),
 (6509, 7445, 940315, 55368, 24, 2307, 'A'),
 (6520, 7485, 940316, 91200, 24, 3800, 'A'),
 (6699, 8330, 940330, 7656, 24, 319, 'A'),
 (6336, 6555, 940331, 212400, 36, 5900, 'A'),
 (6202, 5891, 940418, 65184, 12, 5432, 'A'),
 (7066, 10131, 940502, 215388, 36, 5983, 'B'),
 (6072, 5256, 940510, 193968, 36, 5388, 'A'),
 (5130, 813, 940511, 24312, 12, 2026, 'A'),
 (6678, 8173, 940520, 80340, 12, 6695, 'A'),
 (6998, 9833, 940521, 97632, 12, 8136, 'A'),
 (6580, 7769, 940522, 196080, 48, 4085, 'A'),
 (6316, 6473, 940526, 76908, 12, 6409, 'B'),
 (6512, 7454, 940526, 63312, 48, 1319, 'A'),
 (6407, 6930, 940526, 25944, 24, 1081, 'A'),
 (5002, 173, 940531, 104808, 12, 8734, 'A'),
 (6647, 8051, 940601, 208320, 48, 4340, 'A'),
 (7137, 10439, 940603, 151560, 24, 6315, 'A'),
 (6903, 9337, 940604, 52128, 24, 2172, 'A'),
 (5918, 4576, 940604, 300204, 36, 8339, 'A'),
 (6642, 8027, 940605, 77640, 12, 6470, 'A'),
 (6210, 5927, 940606, 87216, 48, 1817, 'B'),
 (6279, 6273, 940607, 87360, 24, 3640, 'A'),
 (7249, 11065, 940612, 69360, 48, 1445, 'A'),
 (6112, 5442, 940617, 174744, 24, 7281, 'B'),
 (6596, 7844, 940625, 91584, 24, 3816, 'A'),
 (5657, 3354, 940705, 4980, 12, 415, 'A'),
 (5533, 2762, 940705, 111384, 24, 4641, 'A'),
 (6232, 6051, 940708, 276660, 36, 7685, 'B'),
 (6888, 9265, 940711, 39576, 12, 3298, 'A'),
 (6066, 5228, 940711, 173808, 24, 7242, 'A'),
 (6004, 4937, 940723, 143904, 24, 5996, 'A'),
 (5126, 790, 940724, 208128, 48, 4336, 'B'),
 (7220, 10911, 940728, 131292, 36, 3647, 'A'),
 (6668, 8129, 940730, 74736, 36, 2076, 'A'),
 (5395, 2176, 940730, 299088, 36, 8308, 'B'),
 (6064, 5215, 940804, 14628, 12, 1219, 'A'),
 (6919, 9403, 940805, 120888, 36, 3358, 'A'),
 (6841, 9104, 940807, 172080, 48, 3585, 'A'),
 (6038, 5120, 940810, 323472, 48, 6739, 'A'),
 (6729, 8519, 940811, 74688, 12, 6224, 'A'),
 (7198, 10788, 940823, 45768, 12, 3814, 'A'),
 (6686, 8260, 940825, 224604, 36, 6239, 'A'),
 (5836, 4259, 940827, 79632, 24, 3318, 'A'),
 (7061, 10105, 940827, 121896, 24, 5079, 'A'),
 (5595, 3050, 940902, 82896, 12, 6908, 'A'),
 (5723, 3674, 940910, 23628, 12, 1969, 'A'),
 (6120, 5481, 940911, 209184, 48, 4358, 'A'),
 (6895, 9307, 940919, 41904, 12, 3492, 'A'),
 (5882, 4427, 940926, 87816, 12, 7318, 'A'),
 (5754, 3834, 940928, 23052, 12, 1921, 'A'),
 (6594, 7824, 940928, 189696, 24, 7904, 'B'),
 (6738, 8566, 940929, 230220, 36, 6395, 'B'),
 (6624, 7965, 940930, 71460, 12, 5955, 'A'),
 (6216, 5952, 940930, 141240, 24, 5885, 'A'),
 (6229, 6040, 941002, 11736, 24, 489, 'A'),
 (5887, 4448, 941003, 192744, 36, 5354, 'A'),

5. Ordenar lo anterior por duración y cantidad

extra: limitar a 20 elementos

In [ ]:
%%sql 
SELECT *
FROM LOAN
WHERE status in ("A", "B")
ORDER BY duration, amount DESC
LIMIT 100

 * sqlite:///data_berka.db
Done.


[(6934, 9494, 970209, 116832, 12, 9736, 'A'),
 (5002, 173, 940531, 104808, 12, 8734, 'A'),
 (4986, 97, 970810, 102876, 12, 8573, 'A'),
 (5375, 2106, 961003, 101520, 12, 8460, 'A'),
 (7253, 11079, 970426, 98304, 12, 8192, 'A'),
 (6998, 9833, 940521, 97632, 12, 8136, 'A'),
 (6612, 7907, 970831, 97392, 12, 8116, 'A'),
 (5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (6995, 9814, 960926, 96168, 12, 8014, 'A'),
 (5464, 2419, 950130, 95808, 12, 7984, 'A'),
 (5643, 3273, 970921, 94512, 12, 7876, 'B'),
 (5377, 2116, 970804, 92400, 12, 7700, 'A'),
 (5519, 2694, 940308, 91632, 12, 7636, 'A'),
 (7046, 10049, 970328, 91632, 12, 7636, 'A'),
 (5868, 4361, 970329, 90156, 12, 7513, 'A'),
 (7277, 11231, 971024, 89280, 12, 7440, 'A'),
 (4996, 132, 961106, 88440, 12, 7370, 'A'),
 (5882, 4427, 940926, 87816, 12, 7318, 'A'),
 (7119, 10361, 970602, 86880, 12, 7240, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (6921, 9422, 961029, 86184, 12, 7182, 'A'),
 (5366, 2062, 961029, 86184, 12, 7182, 'A'),
 (6150, 5622, 941102, 83016, 12, 6918, 'A'),
 (5595, 3050, 940902, 82896, 12, 6908, 'A'),
 (6678, 8173, 940520, 80340, 12, 6695, 'A'),
 (6059, 5196, 971228, 79824, 12, 6652, 'A'),
 (5285, 1603, 940206, 78936, 12, 6578, 'A'),
 (7051, 10068, 970728, 77808, 12, 6484, 'A'),
 (6642, 8027, 940605, 77640, 12, 6470, 'A'),
 (5117, 718, 970820, 76944, 12, 6412, 'A'),
 (6316, 6473, 940526, 76908, 12, 6409, 'B'),
 (5739, 3747, 950102, 76380, 12, 6365, 'A'),
 (5992, 4863, 970318, 76080, 12, 6340, 'A'),
 (6729, 8519, 940811, 74688, 12, 6224, 'A'),
 (5878, 4415, 970401, 74040, 12, 6170, 'A'),
 (6599, 7859, 961119, 72120, 12, 6010, 'A'),
 (6624, 7965, 940930, 71460, 12, 5955, 'A'),
 (5352, 1993, 950415, 68400, 12, 5700, 'B'),
 (5638, 3249, 970827, 68124, 12, 5677, 'A'),
 (6598, 7857, 950118, 68112, 12, 5676, 'A'),
 (6753, 8688, 970320, 67980, 12, 5665, 'A'),
 (5682, 3455, 970307, 67464, 12, 5622, 'A'),
 (7201, 10807, 970307, 67464, 12, 5622, 'A'),
 (6202, 5891, 940418, 65184, 12, 5432, 'A'),
 (5750, 3804, 960208, 63252, 12, 5271, 'A'),
 (5388, 2153, 961128, 61560, 12, 5130, 'A'),
 (5338, 1888, 970830, 57360, 12, 4780, 'B'),
 (7305, 11359, 960806, 54024, 12, 4502, 'A'),
 (5398, 2187, 971104, 53472, 12, 4456, 'A'),
 (5134, 825, 970202, 53076, 12, 4423, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (5856, 4320, 970809, 51360, 12, 4280, 'B'),
 (6650, 8073, 940217, 49320, 12, 4110, 'B'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7198, 10788, 940823, 45768, 12, 3814, 'A'),
 (7199, 10789, 971209, 45768, 12, 3814, 'A'),
 (7264, 11141, 971007, 44940, 12, 3745, 'A'),
 (6137, 5574, 960410, 44628, 12, 3719, 'A'),
 (6144, 5593, 960809, 44376, 12, 3698, 'A'),
 (6315, 6472, 950206, 43164, 12, 3597, 'A'),
 (6621, 7957, 970614, 42900, 12, 3575, 'A'),
 (6727, 8505, 971210, 42840, 12, 3570, 'A'),
 (6535, 7546, 951102, 41952, 12, 3496, 'A'),
 (6895, 9307, 940919, 41904, 12, 3492, 'A'),
 (6888, 9265, 940711, 39576, 12, 3298, 'A'),
 (6887, 9263, 970508, 39576, 12, 3298, 'A'),
 (5634, 3229, 970501, 39024, 12, 3252, 'A'),
 (5358, 2018, 970121, 38520, 12, 3210, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6362, 6738, 951031, 37908, 12, 3159, 'A'),
 (6306, 6425, 941220, 36684, 12, 3057, 'A'),
 (6929, 9472, 950508, 36396, 12, 3033, 'A'),
 (7008, 9897, 950922, 36168, 12, 3014, 'A'),
 (6310, 6440, 970617, 34572, 12, 2881, 'A'),
 (5932, 4626, 970618, 33348, 12, 2779, 'A'),
 (6758, 8742, 970729, 33144, 12, 2762, 'A'),
 (5576, 2962, 941221, 31248, 12, 2604, 'A'),
 (6501, 7418, 960429, 30276, 12, 2523, 'B'),
 (4961, 19, 960429, 30276, 12, 2523, 'B'),
 (6502, 7424, 971208, 30276, 12, 2523, 'A'),
 (4962, 25, 971208, 30276, 12, 2523, 'A'),
 (6865, 9199, 960911, 29448, 12, 2454, 'B'),
 (6578, 7765, 951211, 27984, 12, 2332, 'A'),
 (5005, 176, 970410, 27456, 12, 2288, 'A'),
 (5479, 2486, 940110, 24516, 12, 2043, 'A'),
 (5130, 813, 940511, 24312, 12, 2026, 'A'),
 (5723, 3674, 940910, 23628, 12, 1969, 'A'),
 (5229, 1329, 970906, 23520, 12, 1960, 'A'),
 (6878, 9242, 970124, 23184, 12, 1932

### ***Cargar el resultado de la sql en un DataFrame de Pandas:***

En la primera línea he importado la libreria. 

En la segunda linea he ejecutado el comando sql. 

En la tercera línea transformo en un DataFrame de pandas y lo descargo en csv. 

In [ ]:
import pandas as pd
df = %sql SELECT *FROM LOAN WHERE status in ("A", "B") ORDER BY duration, amount DESC
df.DataFrame(). to_csv("first_sql.csv")

 * sqlite:///data_berka.db
Done.


 Ex. 1 

> Seleccionar fecha, cantidad, duración y pagos de aquellos préstamos que han sido terminados ordenados de mayor a menor cantidad de dinero y que tengan una duración inferior a un año. 

In [ ]:
%%sql 
SELECT date, amount, duration, payments
FROM LOAN
WHERE status in ("A", "B")  AND duration > 12
ORDER  by amount DESC
LIMIT 100

 * sqlite:///data_berka.db
Done.


[(931201, 464520, 60, 7742),
 (940810, 323472, 48, 6739),
 (940604, 300204, 36, 8339),
 (940730, 299088, 36, 8308),
 (950103, 288468, 36, 8013),
 (940708, 276660, 36, 7685),
 (930906, 274740, 60, 4579),
 (950108, 270648, 36, 7518),
 (950623, 265320, 36, 7370),
 (951028, 260028, 36, 7223),
 (931213, 259740, 60, 4329),
 (950428, 248148, 36, 6893),
 (951020, 231696, 36, 6436),
 (940929, 230220, 36, 6395),
 (941012, 227880, 36, 6330),
 (940825, 224604, 36, 6239),
 (951204, 217152, 36, 6032),
 (940502, 215388, 36, 5983),
 (940331, 212400, 36, 5900),
 (940911, 209184, 48, 4358),
 (940601, 208320, 48, 4340),
 (940724, 208128, 48, 4336),
 (941204, 205200, 36, 5700),
 (950624, 198240, 24, 8260),
 (940223, 197748, 36, 5493),
 (940522, 196080, 48, 4085),
 (940510, 193968, 36, 5388),
 (941003, 192744, 36, 5354),
 (941112, 192552, 24, 8023),
 (940928, 189696, 24, 7904),
 (960101, 189696, 24, 7904),
 (950501, 187224, 24, 7801),
 (960728, 187104, 24, 7796),
 (940301, 182628, 36, 5073),
 (930924, 174744, 24, 7281),
 (940617, 174744, 24, 7281),
 (940711, 173808, 24, 7242),
 (940807, 172080, 48, 3585),
 (941030, 170256, 24, 7094),
 (950304, 168984, 24, 7041),
 (930711, 165960, 36, 4610),
 (960502, 165960, 24, 6915),
 (941119, 164052, 36, 4557),
 (950401, 156672, 36, 4352),
 (960715, 155880, 24, 6495),
 (960804, 155640, 24, 6485),
 (931013, 154416, 48, 3217),
 (940603, 151560, 24, 6315),
 (940207, 149040, 48, 3105),
 (940723, 143904, 24, 5996),
 (960822, 141648, 24, 5902),
 (940930, 141240, 24, 5885),
 (941106, 140400, 36, 3900),
 (950613, 139536, 24, 5814),
 (951128, 137904, 24, 5746),
 (950319, 134784, 24, 5616),
 (940728, 131292, 36, 3647),
 (961227, 129408, 24, 5392),
 (930728, 127080, 60, 2118),
 (950413, 126600, 24, 5275),
 (941103, 123696, 48, 2577),
 (961211, 123408, 24, 5142),
 (940827, 121896, 24, 5079),
 (940805, 120888, 36, 3358),
 (931104, 117024, 24, 4876),
 (941221, 115992, 36, 3222),
 (951122, 113796, 36, 3161),
 (940705, 111384, 24, 4641),
 (961111, 110136, 24, 4589),
 (961109, 110112, 24, 4588),
 (940131, 108144, 36, 3004),
 (950811, 107640, 24, 4485),
 (930803, 105804, 36, 2939),
 (950411, 104712, 24, 4363),
 (951212, 104712, 24, 4363),
 (951104, 99336, 24, 4139),
 (950710, 98592, 24, 4108),
 (941028, 98184, 24, 4091),
 (950223, 97224, 24, 4051),
 (931208, 95400, 36, 2650),
 (931208, 93888, 36, 2608),
 (961020, 93600, 24, 3900),
 (951212, 92976, 24, 3874),
 (940625, 91584, 24, 3816),
 (940316, 91200, 24, 3800),
 (960822, 91152, 24, 3798),
 (960119, 89880, 24, 3745),
 (930913, 87840, 24, 3660),
 (940607, 87360, 24, 3640),
 (940606, 87216, 48, 1817),
 (960922, 84120, 24, 3505),
 (940105, 80952, 24, 3373),
 (940105, 80952, 24, 3373),
 (940827, 79632, 24, 3318),
 (931122, 79608, 24, 3317),
 (950311, 76680, 24, 3195),
 (950111, 76416, 24, 3184),
 (931223, 75624, 24, 3151),
 (940730, 74736, 36, 2076),
 (951031, 73836, 36, 2051)]

¿Cuántos registros cumplen esa condicion?

In [ ]:
%%sql
SELECT count (*)
FROM LOAN
WHERE status in ("A", "B") AND duration > 12

 * sqlite:///data_berka.db
Done.


[(131,)]

Una vez aplicamos el comando, nos sale que aquellos que pertenecen a esa tabla y que se encuentran en estado A o B y con una duración superior a 12 meses son 131. 

¿Cuántos prestamos hay en ejecución (running) con duración entre 24 y 48 meses

In [ ]:
%%sql
SELECT count (*)
FROM LOAN
WHERE status in ("C", "D") AND duration  between 24 AND 48

 * sqlite:///data_berka.db
Done.


[(279,)]

# Group By

### 7. **¿Cuántos préstamos hay por cada estado y cual es la media de la cantidad de dinero prestado?**

In [ ]:
%%sql 
SELECT count(*)
FROM LOAN

 * sqlite:///data_berka.db
Done.


[(682,)]

Al aplicar el comando obtenemos que hay 682 préstamos. 

7.1 **¿CUÁNTOS PRESTAMOS HAY POR CADA ESTADO?**

In [ ]:
%%sql
SELECT status, count (*) as total
FROM LOAN
GROUP BY status
ORDER BY total DESC

 * sqlite:///data_berka.db
Done.


[('C', 403), ('A', 203), ('D', 45), ('B', 31)]

Con el comando anterior me ha seleccionado los préstamos por tipos de estados (con el as total le indico que me nombre esa columna como total), me lo indica por grupos de estado y me ordena el total de forma descendiente. 


7.2 **¿Cual es la media de la cantidad de dinero prestado?**

In [ ]:
%%sql
SELECT status, count (*) as total, avg (amount)
FROM LOAN
GROUP BY status
ORDER BY total DESC

 * sqlite:///data_berka.db
Done.


[('C', 403, 171410.3523573201),
 ('A', 203, 91641.45812807881),
 ('D', 45, 249284.53333333333),
 ('B', 31, 140720.90322580645)]

La última línea AVG es la que me indica la media de cada estado. 

7.3  ***Sácame cuáles son los que tienen una duración igual o mayor a 24. ***

In [ ]:
%%sql
SELECT status, count (*) as total, avg (amount)
FROM LOAN
WHERE duration >= 24
GROUP BY status
ORDER BY total DESC

 * sqlite:///data_berka.db
Done.


[('C', 376, 180343.1489361702),
 ('A', 110, 122425.2),
 ('D', 44, 254127.27272727274),
 ('B', 21, 179909.14285714287)]

La última línea AVG es la que me indica la media de cada estado. 

7.4 **Ahora solo quiero que me muestre los grupos que tienen mas de 25 prestamos concedidos.** 

where - se filtran los registros antes de agruparlos
habing - se filtran los grupos tras obtenerlos

In [ ]:
%%sql
SELECT status, count (*) as total, avg (amount)
FROM LOAN
WHERE duration >= 24
GROUP BY status
HAVING total > 25
ORDER BY total DESC

 * sqlite:///data_berka.db
Done.


[('C', 376, 180343.1489361702),
 ('A', 110, 122425.2),
 ('D', 44, 254127.27272727274)]

La última línea AVG es la que me indica la media de cada estado. 

7.5. ***Por cada Región, quiero contar el número de ciudades y obtener el total del número de habitantes de la región. ***




In [ ]:
%%sql
SELECT "A3", count (A2) no_districts, sum (A9) no_cities, sum (A4) total_hab
FROM DISTRICT
GROUP BY  "A3"

 * sqlite:///data_berka.db
Done.


[('Prague', 1, 1, 1204953),
 ('central Bohemia', 12, 66, 1105234),
 ('east Bohemia', 11, 86, 1234781),
 ('north Bohemia', 10, 70, 1178977),
 ('north Moravia', 11, 60, 1970302),
 ('south Bohemia', 8, 48, 700595),
 ('south Moravia', 14, 82, 2054989),
 ('west Bohemia', 10, 69, 859306)]

7.6 **Además quiero que añadas un filtro en el que no considere la región de Praga**

In [ ]:
%%sql
SELECT "A3", count (A2) no_districts, sum (A9) no_cities, sum (A4) total_hab
FROM DISTRICT
WHERE NOT A3 = "Prague"
GROUP BY  "A3"

 * sqlite:///data_berka.db
Done.


[('central Bohemia', 12, 66, 1105234),
 ('east Bohemia', 11, 86, 1234781),
 ('north Bohemia', 10, 70, 1178977),
 ('north Moravia', 11, 60, 1970302),
 ('south Bohemia', 8, 48, 700595),
 ('south Moravia', 14, 82, 2054989),
 ('west Bohemia', 10, 69, 859306)]

El having es un resultado sobre los grupos ya hechos. 

Habría varias formas de hacerlo, lo podemos hacer con el having o con el where not. 

# JOIN

Con join hago consultas de distintas tablas. 

### 8. **¿Cuántas cuentas tenemos en la base de datos?**



In [ ]:
%%sql
SELECT count(*)
FROM ACCOUNT

 * sqlite:///data_berka.db
Done.


[(4500,)]

Esto nos indica que tenemos 4500 cuentas. 

8.1 **Ahora sacamos todos los campos que hay dentro de esas tablas, es decir, todo lo que las componen.** 

In [ ]:
%%sql
SELECT loan_id, l.date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN l, ACCOUNT a
WHERE l.account_id = a.account_id
AND status in ("A", "B")
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(5314, 930705, 96396, 12, 8033, 'weekly', 930322),
 (5316, 930711, 165960, 36, 4610, 'monthly', 930213),
 (6863, 930728, 127080, 60, 2118, 'monthly', 930208),
 (5325, 930803, 105804, 36, 2939, 'monthly', 930130),
 (7240, 930906, 274740, 60, 4579, 'weekly', 930214),
 (6687, 930913, 87840, 24, 3660, 'monthly', 930512),
 (7284, 930915, 52788, 12, 4399, 'monthly', 930114),
 (6111, 930924, 174744, 24, 7281, 'monthly', 930414),
 (7235, 931013, 154416, 48, 3217, 'weekly', 930420),
 (5997, 931104, 117024, 24, 4876, 'monthly', 930620)]

Como la columna account la tengo tanto en la tabla LOAN como en ACCOUNT, se lo indico poniendoles un igual. 

8.3 **Úneme las tablas de "ACCOUNT" y "LOAN"**

In [ ]:
%%sql
SELECT *
FROM LOAN l, ACCOUNT a
WHERE l.account_id = a.account_id
LIMIT 50

 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B', 1787, 30, 'weekly', 930322),
 (5316, 1801, 930711, 165960, 36, 4610, 'A', 1801, 46, 'monthly', 930213),
 (6863, 9188, 930728, 127080, 60, 2118, 'A', 9188, 45, 'monthly', 930208),
 (5325, 1843, 930803, 105804, 36, 2939, 'A', 1843, 12, 'monthly', 930130),
 (7240, 11013, 930906, 274740, 60, 4579, 'A', 11013, 1, 'weekly', 930214),
 (6687, 8261, 930913, 87840, 24, 3660, 'A', 8261, 13, 'monthly', 930512),
 (7284, 11265, 930915, 52788, 12, 4399, 'A', 11265, 15, 'monthly', 930114),
 (6111, 5428, 930924, 174744, 24, 7281, 'B', 5428, 24, 'monthly', 930414),
 (7235, 10973, 931013, 154416, 48, 3217, 'A', 10973, 18, 'weekly', 930420),
 (5997, 4894, 931104, 117024, 24, 4876, 'A', 4894, 14, 'monthly', 930620),
 (7121, 10364, 931110, 21924, 36, 609, 'A', 10364, 55, 'monthly', 930117),
 (6077, 5270, 931122, 79608, 24, 3317, 'A', 5270, 44, 'monthly', 930113),
 (6228, 6034, 931201, 464520, 60, 7742, 'B', 6034, 6, 'weekly', 930319),
 (6356, 6701, 931208, 95400, 36, 2650, 'A', 6701, 49, 'weekly', 930315),
 (5523, 2705, 931208, 93888, 36, 2608, 'A', 2705, 63, 'after_tr', 930705),
 (6456, 7123, 931209, 47016, 12, 3918, 'A', 7123, 64, 'after_tr', 930414),
 (7104, 10320, 931213, 259740, 60, 4329, 'A', 10320, 68, 'monthly', 930711),
 (6820, 9034, 931216, 38148, 12, 3179, 'A', 9034, 1, 'monthly', 930628),
 (6876, 9236, 931221, 86616, 12, 7218, 'A', 9236, 11, 'monthly', 930818),
 (7097, 10266, 931223, 75624, 24, 3151, 'B', 10266, 54, 'monthly', 930610),
 (4959, 2, 940105, 80952, 24, 3373, 'A', 2, 1, 'monthly', 930226),
 (6499, 7401, 940105, 80952, 24, 3373, 'A', 7401, 1, 'monthly', 930226),
 (5479, 2486, 940110, 24516, 12, 2043, 'A', 2486, 1, 'monthly', 930227),
 (5170, 1071, 940120, 253200, 60, 4220, 'C', 1071, 68, 'weekly', 930510),
 (6721, 8475, 940121, 21072, 24, 878, 'A', 8475, 4, 'monthly', 930908),
 (7259, 11111, 940131, 108144, 36, 3004, 'A', 11111, 1, 'monthly', 930520),
 (5285, 1603, 940206, 78936, 12, 6578, 'A', 1603, 1, 'monthly', 930509),
 (5189, 1166, 940207, 149040, 48, 3105, 'A', 1166, 11, 'monthly', 930421),
 (6650, 8073, 940217, 49320, 12, 4110, 'B', 8073, 24, 'after_tr', 930903),
 (7226, 10940, 940223, 197748, 36, 5493, 'A', 10940, 50, 'monthly', 930913),
 (6087, 5313, 940227, 300660, 60, 5011, 'C', 5313, 40, 'monthly', 930429),
 (7262, 11135, 940301, 182628, 36, 5073, 'A', 11135, 75, 'monthly', 931010),
 (5519, 2694, 940308, 91632, 12, 7636, 'A', 2694, 7, 'monthly', 930308),
 (6577, 7753, 940311, 51696, 24, 2154, 'A', 7753, 74, 'monthly', 930208),
 (6509, 7445, 940315, 55368, 24, 2307, 'A', 7445, 64, 'monthly', 930318),
 (6520, 7485, 940316, 91200, 24, 3800, 'A', 7485, 39, 'monthly', 930728),
 (6699, 8330, 940330, 7656, 24, 319, 'A', 8330, 48, 'monthly', 930328),
 (6336, 6555, 940331, 212400, 36, 5900, 'A', 6555, 38, 'weekly', 930305),
 (7055, 10079, 940406, 167100, 60, 2785, 'C', 10079, 62, 'monthly', 930506),
 (6103, 5385, 940407, 149340, 60, 2489, 'C', 5385, 44, 'weekly', 930521),
 (6202, 5891, 940418, 65184, 12, 5432, 'A', 5891, 54, 'monthly', 930125),
 (7066, 10131, 940502, 215388, 36, 5983, 'B', 10131, 1, 'monthly', 931127),
 (6696, 8321, 940509, 89040, 60, 1484, 'C', 8321, 77, 'monthly', 930412),
 (6072, 5256, 940510, 193968, 36, 5388, 'A', 5256, 72, 'after_tr', 931025),
 (5130, 813, 940511, 24312, 12, 2026, 'A', 813, 16, 'monthly', 930210),
 (5568, 2933, 940517, 272520, 60, 4542, 'C', 2933, 10, 'monthly', 930618),
 (6678, 8173, 940520, 80340, 12, 6695, 'A', 8173, 11, 'monthly', 930705),
 (6998, 9833, 940521, 97632, 12, 8136, 'A', 9833, 20, 'monthly', 930724),
 (6580, 7769, 940522, 196080, 48, 4085, 'A', 7769, 23, 'monthly', 931003),
 (6316, 6473, 940526, 76908, 12, 6409, 'B', 6473, 12, 'monthly', 930126)]

8.4 **¿Cómo hago para vincular también la tabla de distric a esta?**

In [ ]:
%%sql
SELECT loan_id, A4, A9, A10, A11, l.date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN l, ACCOUNT a, DISTRICT d
WHERE l.account_id = a.account_id
AND a.district_id = d.A1
AND status in ("A", "B")
ORDER BY amount DESC, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(6228, 77963, 4, 51.5, 8546, 931201, 464520, 60, 7742, 'weekly', 930319),
 (6038, 1204953, 1, 100.0, 12541, 940810, 323472, 48, 6739, 'monthly', 930621),
 (5918, 67298, 7, 43.8, 8561, 940604, 300204, 36, 8339, 'monthly', 930605),
 (5395, 106054, 6, 63.1, 8110, 940730, 299088, 36, 8308, 'monthly', 930606),
 (6526, 86513, 5, 50.5, 8288, 950103, 288468, 36, 8013, 'monthly', 940829),
 (6232, 124605, 8, 46.0, 8772, 940708, 276660, 36, 7685, 'monthly', 930828),
 (7240, 1204953, 1, 100.0, 12541, 930906, 274740, 60, 4579, 'weekly', 930214),
 (6244, 1204953, 1, 100.0, 12541, 950108, 270648, 36, 7518, 'weekly', 940307),
 (6725, 51428, 4, 52.7, 8402, 950623, 265320, 36, 7370, 'monthly', 930918),
 (7176, 75637, 7, 36.5, 9622, 951028, 260028, 36, 7223, 'monthly', 931208)]

8.5 **Quiero añadir el año de nacimiento del cliente a estos datos.**

In [ ]:
%%sql
SELECT loan_id, A4, A9, A10, A11,c.birth_number, l.date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN l, ACCOUNT a, DISTRICT d, DISPOSITION dp, CLIENT c
WHERE l.account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id -- join disposition y account
AND dp.client_id = c.client_id  -- join disposition y client
AND status in ("A", "B")
ORDER BY amount DESC, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(6228, 77963, 4, 51.5, 8546, '406202', 931201, 464520, 60, 7742, 'weekly', 930319),
 (6038, 1204953, 1, 100.0, 12541, '406015', 940810, 323472, 48, 6739, 'monthly', 930621),
 (5918, 67298, 7, 43.8, 8561, '640313', 940604, 300204, 36, 8339, 'monthly', 930605),
 (5395, 106054, 6, 63.1, 8110, '666210', 940730, 299088, 36, 8308, 'monthly', 930606),
 (6526, 86513, 5, 50.5, 8288, '586026', 950103, 288468, 36, 8013, 'monthly', 940829),
 (6232, 124605, 8, 46.0, 8772, '490518', 940708, 276660, 36, 7685, 'monthly', 930828),
 (7240, 1204953, 1, 100.0, 12541, '780907', 930906, 274740, 60, 4579, 'weekly', 930214),
 (6244, 1204953, 1, 100.0, 12541, '371104', 950108, 270648, 36, 7518, 'weekly', 940307),
 (6725, 51428, 4, 52.7, 8402, '485828', 950623, 265320, 36, 7370, 'monthly', 930918),
 (6725, 51428, 4, 52.7, 8402, '501204', 950623, 265320, 36, 7370, 'monthly', 930918)]

Con disposition hacemos una tabla puente, ya que es la única con la que tanto cliente como account tienen las dos relación, para poder unir account con client y con eso poder relacionar el cumpleaños con la cuenta de los clientes. 

8.6 ***Ahora vamos a filtrar por el tipo. ***

In [ ]:
%%sql
SELECT count(*) --loan_id, A4, A9, A10, A11,c.birth_number, l.date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN l, ACCOUNT a, DISTRICT d, DISPOSITION dp, CLIENT c
WHERE l.account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id -- join disposition y account
AND dp.client_id = c.client_id  -- join disposition y client
AND dp.type = "OWNER"
--AND status in ("A", "B")
--ORDER BY amount DESC, a.date DESC
--LIMIT 10

 * sqlite:///data_berka.db
Done.


[(682,)]

Esto nos ha indicado que hay 682 propietarios de cuentas. 

8.7 **Ahora vamos a sacar el año de nacimiento**

In [ ]:
%%sql
SELECT loan_id, A4, A9, A10, A11,c.birth_number, SUBSTR (c.birth_number, 0,3),l.date loan_date, amount, duration, payments, frequency, a.date acc_date
FROM LOAN l, ACCOUNT a, DISTRICT d, DISPOSITION dp, CLIENT c
WHERE l.account_id = a.account_id
AND a.district_id = d.A1
AND dp.account_id = a.account_id -- join disposition y account
AND dp.client_id = c.client_id  -- join disposition y client
AND dp.type = "OWNER"
AND status in ("A", "B")
ORDER BY amount DESC, a.date DESC
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(6228, 77963, 4, 51.5, 8546, '406202', '40', 931201, 464520, 60, 7742, 'weekly', 930319),
 (6038, 1204953, 1, 100.0, 12541, '406015', '40', 940810, 323472, 48, 6739, 'monthly', 930621),
 (5918, 67298, 7, 43.8, 8561, '640313', '64', 940604, 300204, 36, 8339, 'monthly', 930605),
 (5395, 106054, 6, 63.1, 8110, '666210', '66', 940730, 299088, 36, 8308, 'monthly', 930606),
 (6526, 86513, 5, 50.5, 8288, '586026', '58', 950103, 288468, 36, 8013, 'monthly', 940829),
 (6232, 124605, 8, 46.0, 8772, '490518', '49', 940708, 276660, 36, 7685, 'monthly', 930828),
 (7240, 1204953, 1, 100.0, 12541, '780907', '78', 930906, 274740, 60, 4579, 'weekly', 930214),
 (6244, 1204953, 1, 100.0, 12541, '371104', '37', 950108, 270648, 36, 7518, 'weekly', 940307),
 (6725, 51428, 4, 52.7, 8402, '485828', '48', 950623, 265320, 36, 7370, 'monthly', 930918),
 (7176, 75637, 7, 36.5, 9622, '761125', '76', 951028, 260028, 36, 7223, 'monthly', 931208)]

La función SUBSTR nos modifica un texto, en este caso le meto ese comando a la fecha de cumpleaños para que me coja los 2 primeros digitos (le indico de 0 a 3 porque al ser un intervalo abierto no me coge las puntas).

En el caso de que estuviera dividido como numero entero tendríamos que dividirlo entre 1000. 

# Subconsultas

### **9. Prestamos cuya cantidad es superior a la media de los clientes de Praga**

- Cantidad media prestada a los cuentas de Praga 

In [ ]:
%%sql
SELECT avg(l.amount)
FROM LOAN l, ACCOUNT a, DISTRICT d
WHERE l.account_id = a.account_id and a.district_id = d.A1
AND d.A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(153957.2857142857,)]

Con este comando, el avg me indica la media que son: 153957 préstamos. 

In [ ]:
%%sql
SELECT count(*)
FROM LOAN l, ACCOUNT a, DISTRICT d
WHERE l.account_id = a.account_id and a.district_id = d.A1
AND d.A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(84,)]

- Consultar los préstamos superiores a esa cantidad

In [ ]:
%%sql
SELECT count(*)
FROM loan
WHERE amount > (SELECT avg(l.amount)
      FROM LOAN l, ACCOUNT a, DISTRICT d
      WHERE l.account_id = a.account_id 
      AND a.district_id = d.A1
      AND d.A3 = "Prague")


 * sqlite:///data_berka.db
Done.


[(277,)]

Obtenemos que los préstamos con una cantidad mayor a la media de la selección indicada es de 277. 

# Repaso de Join/Group By

## Objetivo 2.1

15 . Clientes de Praga con préstamos según su estado

- ¿Cuántos clientes hay en Praga?
- ¿Cuántas cuentas tienen asociados clientes en Praga?
- ¿cuántas de esas cuentas tienen asociados un préstamo?
- Necesito saber cuántos hay por cada estado (*status*)


In [ ]:
%%sql
SELECT count(*)
FROM CLIENT, DISTRICT
WHERE district_id = A1
and A3 in ("Prague")

 * sqlite:///data_berka.db
Done.


[(663,)]

In [ ]:
%%sql
SELECT *
FROM ACCOUNT 
LIMIT 10

 * sqlite:///data_berka.db
Done.


[(576, 55, 'monthly', 930101),
 (3818, 74, 'monthly', 930101),
 (704, 55, 'monthly', 930101),
 (2378, 16, 'monthly', 930101),
 (2632, 24, 'monthly', 930102),
 (1972, 77, 'monthly', 930102),
 (1539, 1, 'after_tr', 930103),
 (793, 47, 'monthly', 930103),
 (2484, 74, 'monthly', 930103),
 (1695, 76, 'monthly', 930103)]

In [ ]:
%%sql
SELECT *
FROM LOAN 
LIMIT 8

 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (6111, 5428, 930924, 174744, 24, 7281, 'B')]

In [ ]:
%%sql
SELECT count(*)
FROM ACCOUNT a, DISTRICT, LOAN l
WHERE a.district_id = A1
and a.account_id = 1.account_id
and A3 in ("Prague")

 * sqlite:///data_berka.db
(sqlite3.OperationalError) unrecognized token: "1.account_id"
[SQL: SELECT count(*)
FROM ACCOUNT a, DISTRICT, LOAN l
WHERE a.district_id = A1
and a.account_id = 1.account_id
and A3 in ("Prague")]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


In [ ]:
%%sql
SELECT status, count(*) as total
FROM ACCOUNT a, DISTRICT, LOAN l
WHERE a.district_id = A1
and a.account_id = l.account_idand A3 in ("Prague")
GROUP BY status
ORDER BY total

 * sqlite:///data_berka.db
(sqlite3.OperationalError) near "A3": syntax error
[SQL: SELECT status, count(*) as total
FROM ACCOUNT a, DISTRICT, LOAN l
WHERE a.district_id = A1
and a.account_id = l.account_idand A3 in ("Prague")
GROUP BY status
ORDER BY total]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


## Ejercicio

> cantidad de dinero medio en transacciones de clientes propietarios de cuentas en Kolin (owned accounts)

In [ ]:
%%sql
SELECT c.client_id, avg (amount) as av
from client c, district d, disposition ds, account a, trans t
WHERE c.district_id = d.A1 and c.client_id = ds.client_id
    and ds.account_id = a.account_id and a.account_id = t.account_id
and d.A2 = "Kolin" and ds.type = "OWNER" 
GROUP BY c.client_id
ORDER BY av DESC
LIMIT 10



 * sqlite:///data_berka.db
Done.


[(5249, 15611.831578947369),
 (3378, 14803.997402597402),
 (1388, 14464.587096774194),
 (6803, 13825.125153374232),
 (8091, 13227.722338204592),
 (3298, 13074.464857881138),
 (1806, 13037.759375),
 (4331, 12644.448051948051),
 (4974, 12023.291743119267),
 (3080, 11784.50460251046)]

# Extra: Left Join

1. ¿Cuántas cuentas hay? 4500
2. ¿Cuántas cuentas hay con préstamos? (inner join) 682
3. Cuentas con los datos de los préstamos si los hay
4. los ordenamos por cantidad


In [ ]:
%%sql 
SELECT a.account_id, a.date, l.amount, l.status, l.payments, l.duration
FROM ACCOUNT a INNER OUTER JOIN LOAN l ON a.account_id = l.account_id
ORDER BY amount DESC




 * sqlite:///data_berka.db
(sqlite3.OperationalError) unknown or unsupported join type: INNER OUTER
[SQL: SELECT a.account_id, a.date, l.amount, l.status, l.payments, l.duration
FROM ACCOUNT a INNER OUTER JOIN LOAN l ON a.account_id = l.account_id
ORDER BY amount DESC]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


# DataSets

---

Dataset sobre las transacciones con Vyskov, extrayendo los años de las diferentes fechas. 

In [ ]:
%%sql dataset << SELECT substr(a.date,0,3) acc_year, substr(t.date, 0, 3) year,amount, balance, k_symbol
FROM ACCOUNT a, TRANS t, DISTRICT d
WHERE a.district_id = d.A1 and d.A2 = "Vyskov"
AND a.account_id = t.account_id


 * sqlite:///data_berka.db
Done.
Returning data to local variable dataset


In [ ]:
dataset.DataFrame().to_csv("dataset.csv")